gluten-core/src/main/scala/org/apache/gluten/utils/PlanUtil.scala

```diff
@@ -70,6 +70,9 @@ object PlanUtil {
       case _: InputAdapter => false
       case _: WholeStageCodegenExec => false
       case r: ReusedExchangeExec => isVanillaColumnarOp(r.child)
+      case f: FileSourceScanExec =>
+        !f.relation.fileFormat.getClass.getSimpleName.equals("ArrowCSVFileFormat") &&
+        f.supportsColumnar
```

Comment:
Would you like to explain this change? Thanks.

Reply:
This will cause add extra c2r and r2c, the function is invoked in fallback to add c2r operator

Reply:
So what if we add a rule to convert `FileSourceScanExec` to `GlutenFileSourceScanExec` (or something like that) that outputs Arrow columnar data by default? Which is reasonable since Gluten is guaranteed compatible with Arrow at arch level. The code can be placed in `gluten-data` and `cpp/core`.

Reply:
If we don't want to copy&paste code from  `FileSourceScanExec`, we can adopt decorator design pattern to wrap a  `FileSourceScanExec` instance inside `ArrowFileSourceScanExec`.

Reply:
Then we should change the inject rule logic, replace FileSourceScanExec to other class. Like https://github.com/apache/incubator-gluten/blob/main/gluten-core/src/main/scala/org/apache/spark/sql/execution/datasources/GlutenWriterColumnarRules.scala,  https://github.com/apache/incubator-gluten/blob/main/gluten-core/src/main/scala/org/apache/spark/sql/execution/datasources/GlutenWriterColumnarRules.scalathis implement looks complex, do we have easier way to inject the rule?

Reply:
We can simply add a validation line to https://github.com/apache/incubator-gluten/blob/main/gluten-core/src/main/scala/org/apache/gluten/extension/columnar/validator/Validators.scala

And add some transformation code to https://github.com/apache/incubator-gluten/blob/c4b4c72732f2d662c6c7b0b41c46d47f77b281d0/gluten-core/src/main/scala/org/apache/gluten/extension/columnar/TransformSingleNode.scala#L289

Could just follow the normal way we transform file scan to file scan transformer. 

It's awkward that we still need to take time to find a way to correctly inject some rules after such long time of development of Gluten. We need some complete refactors against the base arch eventually...